src/libmtl/include/libmtl/MTLFormula.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+#include "Interval.h"
+
+#include <algorithm>
+#include <cassert>
+#include <optional>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace logic {
+
+using TimePoint    = int;
+using TimeInterval = ::arithmetic::Interval<TimePoint>;
+
+class MTLFormula;
+
+/**
+ * @brief Strong typing of atomic propositions, allows simplified construction from Boolean values
+ * as well.
+ *
+ */
+struct AtomicProposition
+{
+	AtomicProposition() = delete;
+
+	AtomicProposition(const std::string &name) : ap_(name)
+	{
+	}
+
+	template <typename Y, typename T = std::enable_if_t<std::is_same<Y, bool>{}>>
+	AtomicProposition(Y b)
+	{
+		ap_ = b ? "true" : "false";
+	}
+
+	std::string ap_;
+};
+
+bool operator==(const AtomicProposition &lhs, const AtomicProposition &rhs);
+
+/**
+ * @brief A (timed-) word which can be validated against a given MTL-formula.
+ *
+ */
+class MTLWord
+{
+public:
+	friend class MTLFormula;
+
+	MTLWord(std::initializer_list<std::pair<std::vector<AtomicProposition>, TimePoint>> a) : word_(a)
+	{
+	}
+
+	bool satisfies_at(const MTLFormula &phi, std::size_t i) const;
+	bool satisfies(const MTLFormula &phi) const;
+
+private:
+	std::vector<std::pair<std::vector<AtomicProposition>, TimePoint>> word_;
+};
+
+enum class LOP { LAND, LOR, LNEG, LUNTIL, AP };
+
+/**
+ * @brief Class representing an MTL-formula with the usual operators.
+ *
+ */
+class MTLFormula
+{
+public:
+	friend class MTLWord;
+	MTLFormula() = delete;
+	MTLFormula(const AtomicProposition &ap);
+
+	template <typename Y, typename T = std::enable_if_t<std::is_same<Y, bool>{}>>
+	MTLFormula(Y b) : ap_(b), operator_(LOP::AP)
+	{
+	}
+
+	/** Boolean AND-operator **/
+	MTLFormula operator&&(const MTLFormula &rhs) const;
+	/** Boolean OR-operator **/
+	MTLFormula operator||(const MTLFormula &rhs) const;
+	/** Boolean negation-operator **/
+	MTLFormula operator!() const;
+	/// timed-until operator (binary)
+	MTLFormula until(const MTLFormula &rhs, const TimeInterval &duration = TimeInterval()) const;
+
+	/// function to test whether a formula consists solely of an atomic proposition
+	bool
+	is_AP() const
+	{
+		return ap_.has_value();
```

Comment:
Shouldn't this check `operator_` instead / additionally?

Reply:
We can do this, true. By construction, `ap_` is only set by the constructor to a value when an atomic proposition is constructed within a MTL-formula